### ISSUE TITLE
Error when deleting a plugin 

### ISSUE THREAD

**Issue by SajEng** [YahnisElsts/plugin-update-checker#138]
I added the Puc_v4_Factory in my plugin for update system but when trying to deleting the plugin i'm getting error which says "Deletion failed: Internal Server Error".
If i remove the  Puc_v4_Factory update checking code from my plugin then it is deleting successfully.

Need your help on this thanks.

**Comment by YahnisElsts** [comment]
Please take a look at the PHP error log. It might have more information about this error.

My guess would be that there's a fatal PHP error somewhere that caused the uninstallation script to crash, but I'm not sure. You're the first person to report this kind of a problem.

**Comment by Mikk3lRo** [comment]
I just tested one of mine. No problems here. Agree that a fatal error is the most likely cause for the error though.

It could be due to how you've added the PUC code though, since it only happens when you have that in there... Posting (some of) your code might help understand the issue.

**Comment by SajEng** [comment]
Hi ,
Thanks for your reply.
I just included the PUC factory right after the comment of the plugin.
Please check the screenshot.
https://www.screencast.com/t/atyQMxgrVhZ
<https://www.google.com/url?q=https%3A%2F%2Fwww.screencast.com%2Ft%2FatyQMxgrVhZ&sa=D&sntz=1&usg=AFQjCNF2O6jfI_JS5K9sUciD8PsmQ6NMwQ>

**Comment by Mikk3lRo** [comment]
It's not fine actually - the uninstall hook makes it fail in combination with PUC.

From the docs:
> The plugin should not run arbitrary code outside of functions, when registering the uninstall hook. In order to run using the hook, the plugin will have to be included, which means that any code laying outside of a function will be run during the uninstall process. The plugin should not hinder the uninstall process.
> If the plugin can not be written without running code within the plugin, then the plugin should create a file named 'uninstall.php' in the base plugin folder. This file will be called, if it exists, during the uninstall process bypassing the uninstall hook. The plugin, when using the 'uninstall.php' should always check for the 'WP_UNINSTALL_PLUGIN' constant, before executing.

...so a workaround would be to move your uninstall code to an external "uninstall.php" file...

I guess we're not even allowed to call an init function outside of a hook ;)

**Comment by Mikk3lRo** [comment]
Minimal example to reproduce:

	<?php
	/*
	Plugin Name: Test uninstall
	Plugin URI: 
	Description: Whatever
	Version: 0.01
	Author: Mikk3lRo
	*/

	register_uninstall_hook(__FILE__, array('testuninstall', 'uninstallhook'));

	require 'vendor/plugin-update-checker-4.2/plugin-update-checker.php';
	$myUpdateChecker = Puc_v4_Factory::buildUpdateChecker(
		'https://bitbucket.org/mikk3lro/testuninstall',
		__FILE__,
		'testuninstall'
	);

	class testuninstall {
		static function uninstallhook() {}
	}

Error log:

	PHP Fatal error:  Class 'Puc_v4p2_Utils' not found in /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/StateStore.php on line 172
	PHP Stack trace:
	PHP   1. {main}() /myhomeweb/wp-admin/admin-ajax.php:0
	PHP   2. do_action() /myhomeweb/wp-admin/admin-ajax.php:91
	PHP   3. WP_Hook->do_action() /myhomeweb/wp-includes/plugin.php:453
	PHP   4. WP_Hook->apply_filters() /myhomeweb/wp-includes/class-wp-hook.php:323
	PHP   5. call_user_func_array:{/myhomeweb/wp-includes/class-wp-hook.php:298}() /myhomeweb/wp-includes/class-wp-hook.php:298
	PHP   6. wp_ajax_delete_plugin() /myhomeweb/wp-includes/class-wp-hook.php:298
	PHP   7. delete_plugins() /myhomeweb/wp-admin/includes/ajax-actions.php:3875
	PHP   8. get_site_transient() /myhomeweb/wp-admin/includes/plugin.php:885
	PHP   9. apply_filters() /myhomeweb/wp-includes/option.php:1630
	PHP  10. WP_Hook->apply_filters() /myhomeweb/wp-includes/plugin.php:203
	PHP  11. call_user_func_array:{/myhomeweb/wp-includes/class-wp-hook.php:300}() /myhomeweb/wp-includes/class-wp-hook.php:300
	PHP  12. Puc_v4p2_UpdateChecker->injectUpdate() /myhomeweb/wp-includes/class-wp-hook.php:300
	PHP  13. Puc_v4p2_Vcs_PluginUpdateChecker->getUpdate() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/UpdateChecker.php:352
	PHP  14. Puc_v4p2_Plugin_UpdateChecker->getUpdate() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/Vcs/PluginUpdateChecker.php:183
	PHP  15. Puc_v4p2_UpdateChecker->getUpdate() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/Plugin/UpdateChecker.php:321
	PHP  16. Puc_v4p2_StateStore->getUpdate() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/UpdateChecker.php:241
	PHP  17. Puc_v4p2_StateStore->lazyLoad() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/StateStore.php:71
	PHP  18. Puc_v4p2_StateStore->load() /myhomeweb/wp-content/plugins/testuninstall/vendor/plugin-update-checker-4.2/Puc/v4p2/StateStore.php:155

**Comment by YahnisElsts** [comment]
Interesting. I also tried to make a minimal example and it uninstalled without any errors:

```php
<?php
/*
Plugin Name: Uninstall Me
Plugin URI: http://example.com/
Description: Try uninstalling this plugin
Version: 1.0
Author: Janis Elsts
*/

register_uninstall_hook(__FILE__, 'WshUninstallationTest::hook_callback');

require 'plugin-update-checker/plugin-update-checker.php';
$ExampleUpdateChecker = Puc_v4_Factory::buildUpdateChecker(
	'http://localhost/nonexistent.json',
	__FILE__,
	'my-fictional-plugin-slug'
);

class WshUninstallationTest {
	public static function hook_callback() {
		error_log(__METHOD__ . ' was executed successfully');
	}
}
```

**Comment by YahnisElsts** [comment]
Oh wait, I spoke too soon. Investigating now.

**Comment by Mikk3lRo** [comment]
It only happens if you've actually checked for updates as far as my testing goes.

Anyway, here's a fix:

	add_action('init', function(){
		require 'vendor/plugin-update-checker-4.2/plugin-update-checker.php';
		$myUpdateChecker = Puc_v4_Factory::buildUpdateChecker(
			'https://bitbucket.org/mikk3lro/testuninstall',
			__FILE__,
			'testuninstall'
		);
	});

Should be fairly trivial to implement as a permanent solution - ie. not do anything outside a hook.

**Comment by YahnisElsts** [comment]
All right, I see what the problem is: one of the update-related filters gets called after the plugin files have already been removed, so the autoloader fails to load the `Puc_v4p2_Utils` class and the whole thing crashes with a fatal error.

Putting the initialisation code in a hook is a suitable workaround for now. Tomorrow, I'll look for a way to fix this in the library itself.

**Comment by Mikk3lRo** [comment]
Well making sure it doesn't actually produce a fatal error if implemented without hooks is probably a good idea... 

But that being said I actually think the documentation should be updated to advice developers to wrap it in the `init` hook anyway (or some other early hook, perhaps `plugins_loaded` if that'll work).

Including it at all in the rare case we're only loading the plugin to uninstall it is almost asking for trouble. And it's somewhat bad practice to run anything outside a hook if it can do it's job inside one anyway.

I've never run into this problem probably *only* because I've always deemed [`register_deactivation_hook`](https://codex.wordpress.org/Function_Reference/register_deactivation_hook) much more appropriate when I needed to clean anything up. Might be worth considering for you too @SajEng ?

**Comment by SajEng** [comment]
@Mikk3lRo  thanks for your help,
@YahnisElsts  bro i created the PUC object in init hook and issue get fixed. 
It seem init hook not get call while deleting a plugin.
However creating PUC object in init hook is valid way to check updates.

thanks

**Comment by YahnisElsts** [comment]
I just committed a patch that should fix this crash - even if the object creation code is not inside a hook.

> I've never run into this problem probably only because I've always deemed register_deactivation_hook much more appropriate when I needed to clean anything up.

Are you saying that your plugins delete their settings and other data upon deactivation? What about users who might want to deactivate it temporarily, like when checking for conflicts?

**Comment by Mikk3lRo** [comment]
I was thinking about temporary files, rewrite rules and that sort of things. But I see your point. I guess my comment was quite biased by the type of projects I deal with myself. Thinking about it there are clearly a lot of things better done in the uninstall hook.

**Comment by YahnisElsts** [comment]
Personally, I prefer to use `uninstall.php` since it's simple to set up (no hooks required) and you don't need to worry about unintentionally running other code. But this approach has its own disadvantages: for example, constants and utility functions defined in other plugin files are not available by default, and you have to be careful about which files you include.

**Comment by YahnisElsts** [comment]
This should be fixed in 4.3.
